### PR TITLE
Add ref counting to meters and histograms

### PR DESCRIPTION
**Pull request by plorenz**

### CODE REVIEW

File: metrics/registry.go [hunk 1]
```diff
@@ -250,3 +278,10 @@ func (registry *registryImpl) Poll() *metrics_pb.MetricsMessage {
 
 	return (*metrics_pb.MetricsMessage)(builder)
 }
+
+type refCounted interface {
```
Comment (dovholuknf) [inline metrics/registry.go:282]: why not put this interface into the util/concurrenz package and make it public?

Comment (dovholuknf) [reply metrics/registry.go:282]: nevermind - it's a type in there already ... somehow that didn't register even though i just looked at it

File: metrics/registry.go [hunk 2]
```diff
@@ -113,45 +119,67 @@ func (registry *registryImpl) FuncGauge(name string, f func() int64) Gauge {
...
+	})
 
-	meter := &meterImpl{
-		Meter: metrics.NewMeter(),
-		dispose: func() {
-			registry.dispose(name)
-		},
+	meter, ok := metric.(Meter)
+	if !ok {
+		panic(fmt.Errorf("metric '%v' already exists and is not a meter. It is a %v", name, reflect.TypeOf(metric).Name()))
```
Comment (dovholuknf) [inline metrics/registry.go:132]: do we really want a panic here?

Comment (plorenz) [reply metrics/registry.go:132]: It should never happen, but if it does happen, it should happen very quickly, as soon as either the controller starts up or a router starts up or a router connects. We don't have any metrics that would be initialized late. So if it passes smoke, it should be good. It's a fast fail for a programming mistake.

**Review by dovholuknf — APPROVED** [review]

**Pull request merged**